test/src/unit-capi-vfs.cc

```diff
@@ -49,850 +51,555 @@
 #include <thread>
 
 using namespace tiledb::test;
+using tiledb::api::test_support::ordinary_vfs;
 
-struct VFSFx {
-  // The unique directory object
-  tiledb::sm::TemporaryLocalDirectory temp_dir_{"tiledb_test_"};
-
-  const std::string HDFS_TEMP_DIR = "hdfs://localhost:9000/tiledb_test/";
-  const std::string S3_PREFIX = "s3://";
-  const std::string S3_BUCKET = S3_PREFIX + random_name("tiledb") + "/";
-  const std::string S3_TEMP_DIR = S3_BUCKET + "tiledb_test/";
-  const std::string AZURE_PREFIX = "azure://";
-  const std::string AZURE_CONTAINER =
-      AZURE_PREFIX + random_name("tiledb") + "/";
-  const std::string AZURE_TEMP_DIR = AZURE_CONTAINER + "tiledb_test/";
-  const std::string FILE_TEMP_DIR =
-#ifndef _WIN32
-      "file://" +
-#endif
-      temp_dir_.path();
-  const std::string MEMFS_TEMP_DIR = std::string("mem://tiledb_test/");
-
-  // TileDB context and vfs
-  tiledb_ctx_t* ctx_;
-  tiledb_vfs_t* vfs_;
-
-  // Functions
-  VFSFx();
-  ~VFSFx();
-  void check_vfs(const std::string& path);
-  void check_write(const std::string& path);
-  void check_move(const std::string& path);
-  void check_copy(const std::string& path);
-  void check_read(const std::string& path);
-  void check_append(const std::string& path);
-  void check_ls(const std::string& path);
-  static std::string random_name(const std::string& prefix);
-};
-
-VFSFx::VFSFx() {
-  create_ctx_and_vfs(&ctx_, &vfs_);
+// The unique directory object
+tiledb::sm::TemporaryLocalDirectory temp_dir_{"tiledb_test_"};
+
+void require_tiledb_ok(capi_return_t rc) {
+  REQUIRE(rc == TILEDB_OK);
 }
 
-VFSFx::~VFSFx() {
-  tiledb_vfs_free(&vfs_);
-  tiledb_ctx_free(&ctx_);
+void require_tiledb_err(capi_return_t rc) {
+  REQUIRE(rc == TILEDB_ERR);
 }
 
-void VFSFx::check_vfs(const std::string& path) {
-  if constexpr (tiledb::sm::filesystem::s3_enabled) {
-    // Check S3 bucket functionality
-    if (path == S3_TEMP_DIR) {
-      int is_bucket = 0;
-      int rc = tiledb_vfs_is_bucket(ctx_, vfs_, S3_BUCKET.c_str(), &is_bucket);
-      REQUIRE(rc == TILEDB_OK);
-      if (is_bucket) {
-        rc = tiledb_vfs_remove_bucket(ctx_, vfs_, S3_BUCKET.c_str());
-        REQUIRE(rc == TILEDB_OK);
-      }
-      rc = tiledb_vfs_is_bucket(ctx_, vfs_, S3_BUCKET.c_str(), &is_bucket);
-      REQUIRE(rc == TILEDB_OK);
-      REQUIRE(!is_bucket);
+int ls_getter(const char* path, void* data) {
+  auto vec = static_cast<std::vector<std::string>*>(data);
+  vec->emplace_back(path);
+  return 1;
+}
 
-      rc = tiledb_vfs_create_bucket(ctx_, vfs_, S3_BUCKET.c_str());
-      REQUIRE(rc == TILEDB_OK);
-      rc = tiledb_vfs_is_bucket(ctx_, vfs_, S3_BUCKET.c_str(), &is_bucket);
-      REQUIRE(is_bucket);
-    }
-  }
+/**
+ * Conditionally allocates and sets filesystem-specific parameters on a config.
+ *
+ * @return tiledb_config_t* A potentially nullptr config.
+ */
+tiledb_config_t* conditionally_initialized_config() {
```

Comment:
If both azure and s3 are enabled, you are going to double alloc the config here... We have mostly the same code in helpers.cc/create_ctx_and_vfs... We can extract the config creation from create_ctx_and_vfs into an object that is similar to ordinary_vfs (I would name it vfs_config). That way the struct also cleans up the invalid config object in case of an exception. I also wouldn't make the config creation optional... That will simplify your test case as well as you won't need the code under "// Allocate config if needed" anymore.